File: python/intermediate/datatypes/scalar/booleans.py
# Basic boolean values
true_value = True
false_value = False
print(f"Basic boolean values: {true_value}, {false_value}")  # True, False

# Boolean as subclass of int
print("\nBooleans as integers:")
print(f"True as integer: {int(True)}")  # 1
print(f"False as integer: {int(False)}")  # 0
print(f"True + True = {True + True}")  # 2
print(f"True * 8 = {True * 8}")  # 8
print(f"False * 100 = {False * 100}")  # 0

# Comparison operators that return booleans
print("\nComparison operators:")
x, y = 5, 10
print(f"x = {x}, y = {y}")
print(f"x == y: {x == y}")  # False
print(f"x != y: {x != y}")  # True
print(f"x < y: {x < y}")  # True
print(f"x > y: {x > y}")  # False
print(f"x <= y: {x <= y}")  # True
print(f"x >= y: {x >= y}")  # False

# Logical operators
print("\nLogical operators:")
a, b = True, False
print(f"a = {a}, b = {b}")
print(f"a and b: {a and b}")  # False
print(f"a or b: {a or b}")  # True
print(f"not a: {not a}")  # False
print(f"not b: {not b}")  # True

# Short-circuit evaluation
print("\nShort-circuit evaluation:")
def true_func():
    print("true_func called")
    return True

def false_func():
    print("false_func called")
    return False

print("False and true_func():")
result = False and true_func()  # true_func not called
print(f"Result: {result}")  # False

print("\nTrue or false_func():")
result = True or false_func()  # false_func not called
print(f"Result: {result}")  # True

# Truth value testing
print("\nTruth value testing:")
print(f"bool(0): {bool(0)}")  # False
print(f"bool(1): {bool(1)}")  # True
print(f"bool(-5): {bool(-5)}")  # True
print(f"bool(''): {bool('')}")  # False
print(f"bool('hello'): {bool('hello')}")  # True
print(f"bool([]): {bool([])}")  # False
print(f"bool([1, 2]): {bool([1, 2])}")  # True
print(f"bool(None): {bool(None)}")  # False

# Objects that define __bool__ or __len__
print("\nCustom truth testing:")
class AlwaysTrue:
    def __bool__(self):
        return True

class AlwaysFalse:
    def __bool__(self):
        return False

class EmptyList:
    def __len__(self):
        return 0  # Empty, so evaluates to False

print(f"bool(AlwaysTrue()): {bool(AlwaysTrue())}")  # True
print(f"bool(AlwaysFalse()): {bool(AlwaysFalse())}")  # False
print(f"bool(EmptyList()): {bool(EmptyList())}")  # False

# Boolean operations in if statements
print("\nBoolean operations in if statements:")
x = 10
if x > 5 and x < 15:
    print(f"x = {x} is between 5 and 15")

name = "Alice"
if name == "Alice" or name == "Bob":
    print(f"Hello, {name}!")

# Boolean operations with collections
print("\nBoolean operations with collections:")
my_list = [1, 2, 3]
print(f"5 in my_list: {5 in my_list}")  # False
print(f"2 in my_list: {2 in my_list}")  # True
print(f"4 not in my_list: {4 not in my_list}")  # True

# Using any() and all()
print("\nUsing any() and all():")
numbers = [2, 4, 6, 8, 9]
print(f"numbers: {numbers}")
print(f"any(n > 5 for n in numbers): {any(n > 5 for n in numbers)}")  # True
print(f"all(n > 5 for n in numbers): {all(n > 5 for n in numbers)}")  # False
print(f"all(n < 10 for n in numbers): {all(n < 10 for n in numbers)}")  # True

# Conditional expressions (ternary operator)
print("\nConditional expressions:")
age = 20
status = "adult" if age >= 18 else "minor"
print(f"Age {age} is {status}")  # adult

# Common boolean patterns
print("\nCommon boolean patterns:")

# Pattern 1: Default values
value = None
result = value or "default"
print(f"value or default: {result}")  # default

# Pattern 2: Early returns
def process_positive(num):
    if not (isinstance(num, (int, float)) and num > 0):
        return "Invalid input"
    return f"Processing {num}"

print(f"process_positive(5): {process_positive(5)}")  # Processing 5
print(f"process_positive(-3): {process_positive(-3)}")  # Invalid input
print(f"process_positive('hello'): {process_positive('hello')}")  # Invalid input

# Pattern 3: Boolean masking with lists
values = [1, 0, 3, 0, 5]
mask = [bool(x) for x in values]
print(f"values: {values}")
print(f"mask: {mask}")  # [True, False, True, False, True]

# Bitwise operations on booleans
print("\nBitwise operations:")
print(f"True & False: {True & False}")  # False (AND)
print(f"True | False: {True | False}")  # True (OR)
print(f"True ^ False: {True ^ False}")  # True (XOR)
print(f"True ^ True: {True ^ True}")  # False (XOR)

File: rust/basic/scalars/booleans.rs
// booleans.rs

fn main() {
    // Boolean type in Rust
    let true_value: bool = true;
    let false_value: bool = false;

    println!("true_value: {}", true_value); // true
    println!("false_value: {}", false_value); // false

    // Type inference works for booleans
    let inferred_bool = true;
    println!("\nInferred boolean: {}", inferred_bool); // true

    // Logical operations
    println!("\nLogical operations:");
    let a = true;
    let b = false;

    // Using the variables to avoid warnings
    println!("a && b: {}", a && b); // false
    println!("a || b: {}", a || b); // true

    println!("\ntrue && true: {}", true && true); // true
    println!("true && false: {}", true && false); // false
    println!("false && true: {}", false && true); // false
    println!("false && false: {}", false && false); // false

    println!("\ntrue || true: {}", true || true); // true
    println!("true || false: {}", true || false); // true
    println!("false || true: {}", false || true); // true
    println!("false || false: {}", false || false); // false

    println!("\n!true: {}", !true); // false
    println!("!false: {}", !false); // true

    // Short-circuit evaluation
    println!("\nShort-circuit evaluation:");

    fn expensive_computation() -> bool {
        println!("Expensive computation called!");
        true
    }

    // The second operand is not evaluated if the first determines the result
    println!("false && expensive_computation():");
    let result = false && expensive_computation(); // expensive_computation() is not called
    println!("Result: {}", result); // false

    println!("\ntrue || expensive_computation():");
    let result = true || expensive_computation(); // expensive_computation() is not called
    println!("Result: {}", result); // true

    println!("\ntrue && expensive_computation():");
    let result = true && expensive_computation(); // expensive_computation() IS called
    println!("Result: {}", result); // true

    // Comparison operations return booleans
    println!("\nComparison operations:");
    println!("5 > 3: {}", 5 > 3); // true
    println!("5 < 3: {}", 5 < 3); // false
    println!("5 == 5: {}", 5 == 5); // true
    println!("5 != 5: {}", 5 != 5); // false
    println!("5 >= 5: {}", 5 >= 5); // true
    println!("5 <= 5: {}", 5 <= 5); // true

    // Booleans in control flow
    println!("\nControl flow with booleans:");
    let condition = true;

    if condition {
        println!("Condition is true");
    } else {
        println!("Condition is false");
    }

    // Using boolean expressions directly
    let age = 25;
    if age >= 18 {
        println!("Adult");
    } else {
        println!("Minor");
    }

    // Boolean methods
    println!("\nBoolean methods:");
    let b = true;

    // The then() method runs a closure if true
    let result = b.then(|| 42);
    println!("true.then(|| 42): {:?}", result); // Some(42)

    let b = false;
    let result = b.then(|| 42);
    println!("false.then(|| 42): {:?}", result); // None

    // The then_some() method returns Some(value) if true
    let b = true;
    let result = b.then_some(42);
    println!("true.then_some(42): {:?}", result); // Some(42)

    let b = false;
    let result = b.then_some(42);
    println!("false.then_some(42): {:?}", result); // None

    // Converting to/from integers
    println!("\nConversion with integers:");
    let true_as_int = true as i32;
    let false_as_int = false as i32;

    println!("true as i32: {}", true_as_int); // 1
    println!("false as i32: {}", false_as_int); // 0

    // Note: You cannot cast integers to bool directly
    // This won't compile: let bool_from_int = 1 as bool;
    // Instead, use comparison:
    let bool_from_int = 1 != 0;
    println!("1 != 0: {}", bool_from_int); // true

    // Pattern matching with booleans
    println!("\nPattern matching:");
    let value = true;

    match value {
        true => println!("It's true!"),
        false => println!("It's false!"),
    }

    // You can also use if let with booleans (though it's unusual)
    if let true = value {
        println!("Value is true");
    }

    // Booleans implement common traits
    println!("\nTrait implementations:");

    // Display trait
    println!("Display: {}", true); // true

    // Debug trait
    println!("Debug: {:?}", false); // false

    // PartialEq and Eq traits (for equality comparison)
    println!("true == true: {}", true == true); // true
    println!("true == false: {}", true == false); // false

    // Ord trait (for ordering)
    println!("false < true: {}", false < true); // true (false is considered less than true)

    // Default trait
    let default_bool: bool = Default::default();
    println!("Default bool: {}", default_bool); // false

    // Using booleans in structs
    #[derive(Debug)]
    struct User {
        name: String,
        is_active: bool,
        is_admin: bool,
    }

    let user = User {
        name: String::from("Alice"),
        is_active: true,
        is_admin: false,
    };

    println!("\nStruct with booleans: {:?}", user);
    // Let's actually access the fields to avoid dead code warnings
    println!("User {} is active: {}", user.name, user.is_active);
    println!("User {} is admin: {}", user.name, user.is_admin);

    // Bitwise operations on booleans
    println!("\nBitwise operations on booleans:");
    println!("true & true: {}", true & true); // true
    println!("true & false: {}", true & false); // false
    println!("true | false: {}", true | false); // true
    println!("true ^ true: {}", true ^ true); // false (XOR)
    println!("true ^ false: {}", true ^ false); // true

    // Note: & and | don't short-circuit like && and ||
    fn side_effect() -> bool {
        println!("Side effect executed!");
        true
    }

    println!("\nDifference between && and &:");
    println!("false && side_effect():");
    let _ = false && side_effect(); // side_effect() is NOT called

    println!("\nfalse & side_effect():");
    let _ = false & side_effect(); // side_effect() IS called

    // Size of bool
    println!("\nSize of bool: {} byte", std::mem::size_of::<bool>()); // 1 byte

    // Booleans in collections
    let bool_array: [bool; 4] = [true, false, true, false];
    println!("\nBoolean array: {:?}", bool_array);

    let bool_vec: Vec<bool> = vec![true, true, false];
    println!("Boolean vector: {:?}", bool_vec);

    // Practical example: form validation
    fn validate_password(password: &str) -> bool {
        let has_minimum_length = password.len() >= 8;
        let has_uppercase = password.chars().any(|c| c.is_uppercase());
        let has_lowercase = password.chars().any(|c| c.is_lowercase());
        let has_digit = password.chars().any(|c| c.is_digit(10));

        has_minimum_length && has_uppercase && has_lowercase && has_digit
    }

    println!("\nPassword validation:");
    println!("'password': {}", validate_password("password")); // false
    println!("'Password123': {}", validate_password("Password123")); // true
}
